### ISSUE TITLE
ReturnnConfig ignores file caching

### ISSUE THREAD

**Issue by albertz** [rwth-i6/i6_core#310]
When you put a `tk.Path(..., cached=True)` object somewhere into your `ReturnnConfig`, it will not use file caching.

Is this expected behavior? At least I did not expect this.

The reason for that:
```python
config = instanciate_delayed(config)
```
This code recursively goes through `config`, and does:
```python
if isinstance(o, DelayedBase):
    o = o.get()
```
Now, `tk.Path` derives from `DelayedBase`, and has this `get`:
```python
def get(self):
    return self.get_path()
```
And `get_path` returns the uncached file path.

So, in the written RETURNN config, you will not have any hashed file paths.

(Btw, in the `RasrConfig`, I think it calls `str(path)` instead, and `Path.__str__` returns `self.get_cached_path()`.)

Another node: I just realize, the RETURNN config writing is also a separate `Task` anyway, which likely runs on a different node, so file caching could not be done directly there. However, the common `file_caching` function is implemented like:
```python
def file_caching(path):
  return f"`cf {path}`"
```
So, it anyway does not return the file path of a cached file, but just this special formatted string, which only RASR can really handle properly. So `Path.__str__` likely would break other tools.

I don't really have a good solution or suggestion currently.

**Comment by albertz** [comment]
If it is expected behavior, I think at least the documentation in `ReturnnConfig` should warn about this.

But then somewhat my open question remains: How do I do caching here? I see that other people often use caching via RETURNN, e.g. `HDFDataset` and others support it directly, or `ExternSprintDataset`, where the RASR config gets written via `RasrConfig`, thus has cached paths. So is this the suggested solution? Implement it on RETURNN side anyway, also wherever else it is currently missing?

**Comment by critias** [comment]
`Path.get` was introduced to have the same interface between `Variable` and `Path`. Not accessing the cached version by default does feel unintuitive to me as well given that `__str__` uses the cached version.

The described problem that it's written in an external file and read on a different node requires of cause support by the used toolkit, like in the given example with RASR, but that isn't possible with most toolkits.

So I don't have any suggestions for the write on one node read on another setup, changing the default behavior of `get` could be considered and would probably be the more intuitive solution (with a warning that the default behavior changed, and an option to switch back to the old behavior or to disable the warning).

Any other opinions?

Side note: I considered a while ago to deprecating the current `tk.Path(..., cached=True)` in favor of something like `CachedPath(tk.Path(...))` to make the difference more explicit and more flexible (this would allow different caching methods for the same file or even support it for writing), but since it wasn't a pressing issue and the current solution worked I didn't do it in the end. Would anybody find this change useful?

**Comment by michelwi** [comment]
I don't understand what is the issue here: returnn does not support (in general) the "cf /path/to/file" file caching method we commonly use, so ignoring it is the correct solution at the moment.

Also the structure of the recipes (writing the config on a different node) does not support file caching where sisyphus would do the copying. So ignoring it is the correct solution.

I propose two solutions to enable general file caching:
A) returnn gets an implementation where for general parameters where a string path is expected also "cf /path/to/file" is allowed and then returnn would do the file caching.
B) during config writing we inspect every tk.Path object if the cached property is set and then serialize it as `cf("/path/to/file")` + make sure that a `cf` function exists in the prolog.

**Comment by michelwi** [comment]
> This is similar to my solution now. I don't do this automatically but rather manually. 

Yes, the user manually taking care of it however they want is of cause also an option.

> It would have been a bit tricky to do automatically, as it would have to go recursively through objects, e.g. DelayedFormat (not just list/tuple/dict, as instanciate_delayed currently is doing it).

I see. But then there is the question of whether or not the addition of `cf {}` would be wanted if a user already is doing some "post processing" of the generated string. I see some use cases where it would break and others where it would be the intended behaviour.
And I see already the arguments "you should have used a tk.Path object with cached=False instead" and "then you can add 'cf' to your DelayedFormat as well".
We can discuss what is the most expected behavior and then change the implementation.
